### PR TITLE
Add ingest and msgpack encoding options to cryo cat

### PR DESCRIPTION
**Pull request by jnwatson**

### CODE REVIEW

File: synapse/lib/iq.py [hunk 1]
```diff
@@ -829,3 +830,19 @@ def istufo(self, obj):
         self.len(2, obj)
         self.isinstance(obj[0], (type(None), str))
         self.isinstance(obj[1], dict)
+
+@contextlib.contextmanager
+def redirectStdin(new_stdin):
```
Comment (vEpiphyte) [inline synapse/lib/iq.py]: lets make this guy part of the SynTest class directly and add an example of use

File: synapse/tests/test_tools_cryo_cat.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
...
+from unittest.mock import Mock
+
+import msgpack
+
 import synapse.cryotank as s_cryotank
 
 import synapse.tools.cryo.cat as s_cryocat
 
 from synapse.tests.common import *
+from synapse.lib.iq import redirectStdin
```
Comment (vEpiphyte) [inline synapse/tests/test_tools_cryo_cat.py]: when this guy becomes part of syntest we can remove this import

File: synapse/tests/test_tools_cryo_cat.py [hunk 3]
```diff
@@ -49,6 +55,46 @@ def test_cryocat(self):
...
+                inp = io.StringIO('{"foo: "bar"}\n[]\n')
+                with redirectStdin(inp):
+                    self.raises(Exception, s_cryocat.main, argv, outp)
+
+                # Happy path msgpack ingest
+                outp = self.getTestOutp()
+                argv = ['--ingest', '--msgpack', '--authfile', authfp, addr]
+                to_ingest1 = s_msgpack.en({'foo': 'bar'})
+                to_ingest2 = s_msgpack.en(['lol', 'brb'])
+                inp = Mock()
```
Comment (vEpiphyte) [inline synapse/tests/test_tools_cryo_cat.py]: solid use of Mock

File: synapse/lib/iq.py (hunk 1 above)
Comment (vEpiphyte) [reply synapse/lib/iq.py]: lets also include examples showing bytes vs text oriented use

File: synapse/tools/cryo/cat.py [hunk 4]
```diff
@@ -55,11 +59,36 @@ def main(argv, outp=s_output.stdout):
 
         return 0
 
-    for item in cryo.slice(path, opts.offset, opts.size, opts.timeout):
-        if opts.jsonl:
-            outp.printf(json.dumps(item[1], sort_keys=True))
+    def except_wrap(it, error_str_func):
```
Comment (vEpiphyte) [inline synapse/tools/cryo/cat.py]: lets move this guy up above main() since he's not using any locals

File: synapse/tools/cryo/cat.py [hunk 5]
```diff
@@ -55,11 +59,36 @@ def main(argv, outp=s_output.stdout):
...
-            outp.printf(pprint.pformat(item))
+            fd = sys.stdin
+            item_it = except_wrap((json.loads(s) for s in fd), lambda x: ('Failure parsing line %d of input' % x))
+        cryo.puts(path, item_it)
+    else:
+        for item in cryo.slice(path, opts.offset, opts.size, opts.timeout):
+            if opts.jsonl:
+                outp.printf(json.dumps(item[1], sort_keys=True))
+            elif opts.msgpack:
+                sys.stdout.write(s_msgpack.en(item))
```
Comment (vEpiphyte) [inline synapse/tools/cryo/cat.py]: lets unpack the item here like we do in the jsonl case

File: synapse/tools/cryo/cat.py [hunk 6]
```diff
@@ -55,11 +59,36 @@ def main(argv, outp=s_output.stdout):
...
+            else:
+                outp.printf(pprint.pformat(item))
```
Comment (vEpiphyte) [inline synapse/tools/cryo/cat.py]: can we smoke in the optional offset removal here?

File: synapse/lib/iq.py (hunk 1 above)
Comment (vertexmc) [reply synapse/lib/iq.py]: +1 moving `redirectStdin` to SynTest

Comment (jnwatson) [reply synapse/lib/iq.py]: ok

File: synapse/tests/test_tools_cryo_cat.py (hunk 2 above)
Comment (jnwatson) [reply synapse/tests/test_tools_cryo_cat.py]: ok

File: synapse/tools/cryo/cat.py (hunk 4 above)
Comment (jnwatson) [reply synapse/tools/cryo/cat.py]: why?

File: synapse/tools/cryo/cat.py (hunk 5 above)
Comment (jnwatson) [reply synapse/tools/cryo/cat.py]: ok

File: synapse/tools/cryo/cat.py (hunk 6 above)
Comment (jnwatson) [reply synapse/tools/cryo/cat.py]: ok

File: synapse/tools/cryo/cat.py (hunk 4 above)
Comment (jnwatson) [reply synapse/tools/cryo/cat.py]: ah, yeah he used to. ok

**Pull request merged**